src/main/kotlin/io/github/gradlenexus/publishplugin/NexusPublishPlugin.kt

```diff
@@ -40,29 +34,60 @@ import org.gradle.kotlin.dsl.withType
 @Suppress("UnstableApiUsage")
 class NexusPublishPlugin : Plugin<Project> {
 
-    companion object {
-        private val serverUrlToStagingRepoUrl = ConcurrentHashMap<URI, URI>()
-    }
-
     override fun apply(project: Project) {
-        project.pluginManager.apply(MavenPublishPlugin::class)
+        require(project == project.rootProject) {
+            "Plugin must be applied to the root project"
+        }
 
-        project.gradle.addBuildListener(object : BuildAdapter() {
-            override fun buildFinished(result: BuildResult) {
-                serverUrlToStagingRepoUrl.clear()
+        val extension = project.extensions.create<NexusPublishExtension>(NexusPublishExtension.NAME, project)
+        configureTasks(project, extension)
+        configureTaskDependenciesForAllProjects(project, extension)
+
+        project.rootProject.plugins.withId("io.codearte.nexus-staging") {
+            val nexusStagingExtension = project.rootProject.the<NexusStagingExtension>()
+
+            extension.packageGroup.set(project.provider {
+                if (nexusStagingExtension.packageGroup.isNullOrBlank()) {
+                    project.group.toString()
+                } else {
+                    nexusStagingExtension.packageGroup
+                }
+            })
+            extension.repositories.configureEach {
+                stagingProfileId.set(project.provider { nexusStagingExtension.stagingProfileId })
+                username.set(project.provider { nexusStagingExtension.username })
+                password.set(project.provider { nexusStagingExtension.password })
             }
-        })
+        }
+    }
 
-        val extension = project.extensions.create<NexusPublishExtension>(NexusPublishExtension.NAME, project)
+    private fun configureTaskDependenciesForAllProjects(rootProject: Project, extension: NexusPublishExtension) {
+        rootProject.afterEvaluate {
+            rootProject.allprojects {
+                val publishingProject = this
+                pluginManager.withPlugin("maven-publish") {
+                    publishingProject.afterEvaluate {
+                        val nexusRepositories = addMavenRepositories(publishingProject, extension)
+                        nexusRepositories.forEach { (nexusRepo, mavenRepo) ->
+                            val initializeTask = rootProject.tasks.withType(InitializeNexusStagingRepository::class)
+                                    .named("initialize${nexusRepo.capitalizedName()}StagingRepository")
+                            val closeTask = rootProject.tasks.withType(CloseNexusStagingRepository::class)
+                                    .named("close${nexusRepo.capitalizedName()}StagingRepository")
+                            val releaseTask = rootProject.tasks.withType(ReleaseNexusStagingRepository::class)
+                                    .named("release${nexusRepo.capitalizedName()}StagingRepository")
```

Comment:
This looks like a loop invariant. Should it be moved out of the loop? (before `allprojects`)

It would probably make sense to use a single `named` call rather than `withType + named`:

```kotlin
rootProject.tasks.named<InitializeNexusStagingRepository>(
    "initialize${nexusRepo.capitalizedName()}StagingRepository"
)
```

Reply:
Ah, the vals can't be moved out of the loop since they reference `nexusRepo`

Reply:
It's not an invariant as it depends on `nexusRepo`. Good catch with the simpler `named<>`. 👍